### ISSUE TITLE
[MP4Box] Decryption of CENC content fails

### ISSUE THREAD

**Issue by vladimir-kazakov** [gpac/gpac#1047]
When I decrypt CENC content, MP4Box either crashes or hangs. This issue affects the usage of all protection schemes defined in the third edition of the CENC standard (cenc, cens, cbc1 and cbcs).

Test content and crypt.xml files can be found here: https://1drv.ms/f/s!Ah-Kf_FqISskkkmTypYhfJTqICJL.

I use the following command to encrypt content:

```
mp4box.exe -crypt "cenc\crypt.xml" "360p-750k.mp4" -out "cenc\360p-750k.mp4"
```

And the following command to decrypt it:

```
mp4box.exe -decrypt "cenc\crypt.xml" "cenc\360p-750k.mp4" -out "cenc\360p-750k-clear.mp4" -logs all@debug -lf "cenc\decrypt.txt"
```

I'm not sure whether logs show anything interesting, but you can find them together with the test content, in folders named after protection schemes (decrypt.txt files).

I also tried debugging this operation in Visual Studio, and it fails at: https://github.com/gpac/gpac/blob/42cd567b3e17b855e701854e0104c5b4b200d19e/src/isomedia/box_funcs.c#L1388
An exception is being thrown saying "read access violation, a->registry was 0xDDDDDDDD".

Speaking of crashes and hangs, I noticed that the crash happens when decrypting cenc, cbc1 and cens content. When I decrypt cbcs content, then MP4Box hangs.

I'm on Windows 10, 1709 (16299.371), using GPAC version 0.7.2-DEV-rev461-g01391a0c-master. Everything is 64-bit.

**Comment by vladimir-kazakov** [comment]
I tried decrypting cbcs content using Shaka Packager, but it failed saying:
```
[0508/102927:INFO:demuxer.cc(88)] Demuxer::Run() on file 'In/360p-750k.mp4'.
[0508/102927:INFO:demuxer.cc(160)] Initialize Demuxer for file 'In/360p-750k.mp4'.
[0508/102927:ERROR:track_run_iterator.cc(238)] Failure while processing: tri.video_description->sinf.info.track_encryption .default_is_protected == 0
[0508/102927:ERROR:mp4_media_parser.cc(634)] Failure while processing: runs_->Init()
[0508/102927:ERROR:packager_main.cc(497)] Packaging Error: 8 (PARSER_FAILURE): Cannot parse media file In/360p-750k.mp4
```

Then I tried creating cbcs content using Shaka Packager (the file is in /ShakaPackager/cbcs) and decrypting it using MP4Box, and it succeeded. The content was decrypted correctly and the video plays as expected. MP4Box, though, showed following warnings (the debug log is in /ShakaPackager/cbcs/decrypt.txt):
```
[iso file] Unknown box type ID32
[iso file] saiz box without flags nor cenc type, assuming cenc
[iso file] saio box without flags nor cenc type, assuming cenc
[iso file] saiz box without flags nor cenc type, assuming cenc
[iso file] saio box without flags nor cenc type, assuming cenc
[iso file] saiz box without flags nor cenc type, assuming cenc
[iso file] saio box without flags nor cenc type, assuming cenc
[iso file] saiz box without flags nor cenc type, assuming cenc
[iso file] saio box without flags nor cenc type, assuming cenc
[iso file] saiz box without flags nor cenc type, assuming cenc
[iso file] saio box without flags nor cenc type, assuming cenc
[iso file] saiz box without flags nor cenc type, assuming cenc
[iso file] saio box without flags nor cenc type, assuming cenc
```

It's interesting that cbcs content is correctly decrypted even considering the fact that the content is assumed to be cenc. Perhaps warnings are wrong?

It seems that MP4Box crashes while decrypting its own content, because it encrypts the content incorrectly. I'll create a separate issue for that.

**Comment by vladimir-kazakov** [comment]
It turned out that Shaka Packager was showing the previously mentioned error, because it doesn't support non-fragmented input. I created the fragmented version by dashing the encrypted file using the following command:
```
mp4box.exe -dash 10000 -frag 10000 -rap -frag-rap -profile dashavc264:onDemand "cbcs\360p-750k.mp4" -out "cbcs\dash\index.mpd"
```
And now the error says that the subsample size is incorrect:
```
[0515/062737:INFO:demuxer.cc(88)] Demuxer::Run() on file 'In/cbcs/360p-750k.mp4'.
[0515/062737:INFO:demuxer.cc(160)] Initialize Demuxer for file 'In/cbcs/360p-750k.mp4'.
[0515/062737:ERROR:track_run_iterator.cc(609)] Incorrect CENC subsample size.
[0515/062737:ERROR:mp4_media_parser.cc(744)] Missing decrypt config.
[0515/062737:ERROR:packager_main.cc(497)] Packaging Error: 8 (PARSER_FAILURE): Cannot parse media file In/cbcs/360p-750k.mp4
```

I guess it points to the issue that I mentioned here: https://github.com/gpac/gpac/issues/1053#issuecomment-387978061

@rbouqueau, @jeanlf, do you have any ideas what could be wrong and how it could be fixed?

**Comment by jeanlf** [comment]
this should now be fixed, could you cross-check ?

**Comment by rbouqueau** [comment]
@vladimir-kazakov Thank you. Do we have content to test these crashes?

**Comment by jeanlf** [comment]
this should now be fixed, close if OK

**Comment by vladimir-kazakov** [comment]
Thanks! Now everything works as expected. Closing this issue.

**Comment by vladimir-kazakov** [comment]
@jeanlf, somewhere after commit 71e5bba56d46c17b1cfa8aca834918f119c92575, the decryption of cbcs content was broken again. The process ends successfully, but the media player shows "green picture" distortions. This happens on GPAC version 0.7.2-DEV-rev531-g09c3a971-master, but everything works fine on GPAC version 0.7.2-DEV-rev477-g71e5bba5-master.

**Comment by jeanlf** [comment]
could you upload a working and non working file ?

**Comment by vladimir-kazakov** [comment]
Uploaded files to https://1drv.ms/f/s!Ah-Kf_FqISskkkmTypYhfJTqICJL.

I noticed that only CBC-based protection schemes (cbc1 and cbcs) are affected. CTR-based protection schemes (cenc and cens) are not affected and are correctly decrypted. In the mentioned link, you can see folders with files for each affected protection scheme. In *-rev477 folders, there are files created by MP4Box from GPAC version 0.7.2-DEV-rev477-g71e5bba5-master. In other folders, files are created by MP4Box from GPAC version 0.7.2-DEV-rev534-g146aba3e-master. Files named *-clear.mp4 are decrypted ones. Other files are encrypted.

**Comment by jeanlf** [comment]
my bad, it looks like I had one commit not pushed. I'll check with latest git if I can see diffs between them but there will likely be some, as the slice header parsing was broken even in rev477 - 
this applies to both cbc1 and cbcs since we by default now leave the slice header in clear.
If you can cross-check on your side if the latest git fixes this that would be great.

**Comment by vladimir-kazakov** [comment]
I tried GPAC version 0.7.2-DEV-rev535-gb052cb21-master, decryption is still broken.

**Comment by jeanlf** [comment]
let's summarize:
- playback OK on safari and android
- GPAC decryption of the content broken
is that it ?

**Comment by vladimir-kazakov** [comment]
Yes, that's correct.

**Comment by vladimir-kazakov** [comment]
Everything works fine after latest encryption fixes, starting from GPAC version 0.7.2-DEV-rev539-gff59dfa0-master. Closing the issue.
